components/script/dom/htmlformelement.rs

```diff
@@ -972,3 +906,72 @@ impl Runnable for PlannedNavigation {
         }
     }
 }
+
+
+// https://html.spec.whatwg.org/multipage/#multipart/form-data-encoding-algorithm
+pub fn encode_multipart_form_data(form_data: &mut Vec<FormDatum>,
+                                  boundary: String, encoding: EncodingRef) -> Vec<u8> {
+    // Step 1
+    let mut result = vec![];
+
+    // Step 2
+    let charset = &*encoding.whatwg_name().unwrap_or("UTF-8");
+
+    // Step 3
+    for entry in form_data.iter_mut() {
+        // 3.1
+        if entry.name == "_charset_" && entry.ty == "hidden" {
+            entry.value = FormDatumValue::String(DOMString::from(charset.clone()));
+        }
+        // TODO: 3.2
+
+        // Step 4
+        // https://tools.ietf.org/html/rfc7578#section-4
+        // NOTE(izgzhen): The encoding here expected by most servers seems different from
+        // what spec says (that it should start with a '\r\n').
+        let mut boundary_bytes = format!("--{}\r\n", boundary).into_bytes();
+        result.append(&mut boundary_bytes);
+        let mut content_disposition = ContentDisposition {
+            disposition: DispositionType::Ext("form-data".to_owned()),
+            parameters: vec![DispositionParam::Ext("name".to_owned(), String::from(entry.name.clone()))]
+        };
+
+        match entry.value {
+            FormDatumValue::String(ref s) => {
+                let mut bytes = format!("Content-Disposition: {}\r\n\r\n{}",
+                                        content_disposition, s).into_bytes();
+                result.append(&mut bytes);
+            }
+            FormDatumValue::File(ref f) => {
+                content_disposition.parameters.push(
+                    DispositionParam::Filename(Charset::Ext(String::from(charset.clone())),
+                                               None,
+                                               f.name().clone().into()));
+                // https://tools.ietf.org/html/rfc7578#section-4.4
+                let content_type = ContentType(f.upcast::<Blob>().Type()
+                                                .parse().unwrap_or(mime!(Text / Plain)));
+                let mut type_bytes = format!("Content-Disposition: {}\r\ncontent-type: {}\r\n\r\n",
```

Comment:
Note I added the `content-type: ` here: https://tools.ietf.org/html/rfc7578#section-4.5